flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/PartiallyFinishedSourcesITCase.java

```diff
@@ -0,0 +1,100 @@
+import org.apache.flink.runtime.operators.lifecycle.validation.TestJobDataFlowValidator;
+import org.apache.flink.test.util.AbstractTestBase;
+import org.apache.flink.testutils.junit.SharedObjects;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+
+import static org.apache.flink.api.common.restartstrategy.RestartStrategies.fixedDelayRestart;
+import static org.apache.flink.runtime.operators.lifecycle.command.TestCommand.FAIL;
+import static org.apache.flink.runtime.operators.lifecycle.command.TestCommand.FINISH_SOURCES;
+import static org.apache.flink.runtime.operators.lifecycle.command.TestCommandQueue.TestCommandRetention.REMOVE_ON_MATCH;
+import static org.apache.flink.runtime.operators.lifecycle.command.TestCommandQueue.TestCommandRetention.RETAIN_ON_MATCH;
+import static org.apache.flink.runtime.operators.lifecycle.command.TestCommandQueue.TestCommandTarget.ALL;
+import static org.apache.flink.runtime.operators.lifecycle.command.TestCommandQueue.TestCommandTarget.forOperatorId;
+import static org.apache.flink.runtime.operators.lifecycle.graph.TestJobBuilders.COMPLEX_GRAPH_BUILDER;
+import static org.apache.flink.runtime.operators.lifecycle.graph.TestJobBuilders.SIMPLE_GRAPH_BUILDER;
+import static org.apache.flink.runtime.operators.lifecycle.validation.TestOperatorLifecycleValidator.checkOperatorsLifecycle;
+import static org.apache.flink.streaming.api.environment.ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH;
+
+/**
+ * A test suite to check that the operator methods are called according to contract when sources are
+ * finishing partially.. The contract was refined in FLIP-147.
+ */
+@RunWith(Parameterized.class)
+public class PartiallyFinishedSourcesITCase extends AbstractTestBase {
+
+    @Rule public final SharedObjects sharedObjects = SharedObjects.create();
+
+    @Parameter(0)
+    public TestingGraphBuilder graphBuilder;
+
+    @Parameter(1)
+    public boolean fullyFinish;
+
+    @Test
+    public void test() throws Exception {
+        TestJobWithDescription testJob =
+                graphBuilder.apply(
+                        sharedObjects,
+                        cfg -> cfg.setBoolean(ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true),
+                        env -> env.setRestartStrategy(fixedDelayRestart(1, 0)));
+
+        TestJobExecutor.submitGraph(testJob.jobGraph)
+                .waitForEvent(CheckpointCompletedEvent.class, testJob.eventQueue)
+                // finish some sources
+                .sendCommand(
+                        testJob.commandQueue,
+                        FINISH_SOURCES,
+                        fullyFinish ? RETAIN_ON_MATCH : REMOVE_ON_MATCH,
+                        forOperatorId(testJob.sources.iterator().next()))
+                .waitForEvent(CheckpointCompletedEvent.class, testJob.eventQueue)
```

Comment:
I think we should wait for:

1. subtask(s) to FINISH after sending finish command
2. wait for one more checkpoint to start
3. wait for one more checkpoint to complete

Reply:
I added something similar in 38d6eae2ccccd8768676472f8806fa405b042340
```
                .waitForSubtasksToFinish(finishingVertexID, fullyFinish)
                // wait for a checkpoint to complete with a finished subtask(s)
                // but skip one checkpoint that might be started before finishing
                .waitForEvent(CheckpointCompletedEvent.class, testJob.eventQueue)
                .waitForEvent(CheckpointCompletedEvent.class, testJob.eventQueue)
```
If on (2) we wait for start, then on (3) there is no guarantee that this is the checkpoint just started (maybe an earlier one, before finishing).